File: csharp/ToolGood.Algorithm/Internals/Functions/String/Function_MID.cs
using System;
using System.Collections.Generic;
using ToolGood.Algorithm.Enums;

namespace ToolGood.Algorithm.Internals.Functions.String
{
	internal sealed class Function_MID : Function_3
	{
		public Function_MID(FunctionBase[] funcs) : base(funcs)
		{
			if (funcs.Length != 3) {
				throw new ArgumentException($"Function '{Name}' requires exactly 3 parameters.");
			}
		}

		public override string Name => "Mid";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args1 = GetText_1(engine, tempParameter);
			if(args1.IsErrorOrNone) { return args1; }
			var args2 = GetNumber_2(engine, tempParameter);
			if(args2.IsErrorOrNone) { return args2; }
			var args3 = GetNumber_3(engine, tempParameter);
			if(args3.IsErrorOrNone) { return args3; }

			var text = args1.TextValue;
			// 数值超出 int 范围时按参数错误处理,避免 OverflowException 外泄
			if (TryGetInt(args2, out int startValue) == false) {
				return ParameterError(2);
			}
			if (TryGetInt(args3, out int length) == false) {
				return ParameterError(3);
			}
			// 用 long 计算,避免 startValue 为 int.MinValue 时相减溢出
			var startIndex = (long)startValue - engine.ExcelIndex;

			if(startIndex < 0) {
				return ParameterError(2);
			}
			if(length < 0) {
				return ParameterError(3);
			}
			if(startIndex == 0 && length >= text.Length) {
				return args1;
			}
			if(startIndex >= text.Length) {
				return Operand.Create(string.Empty);
			}
			// 用减法比较,避免 startIndex + length 溢出为负数
			if(length > text.Length - startIndex) {
				length = text.Length - (int)startIndex;
			}
			return Operand.Create(text.Substring((int)startIndex, length));
		}
		public override OperandType GetResultType()
		{
			return OperandType.TEXT;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			func1.GetParameterTypes(noneEngine, result, OperandType.TEXT);
			func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
			func3.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
		}
	}

}

File: java/src/main/java/toolgood/algorithm/internals/functions/string/Function_MID.java
package toolgood.algorithm.internals.functions.string;

import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_3;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_MID extends Function_3 {
    public Function_MID(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "Mid";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        Operand args1 = GetText_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) { return args1; }
        Operand args2 = GetNumber_2(engine, tempParameter);
        if (args2.IsErrorOrNone()) { return args2; }
        Operand args3 = GetNumber_3(engine, tempParameter);
        if (args3.IsErrorOrNone()) { return args3; }

        String text = args1.TextValue();
        int startIndex = args2.IntValue() - engine.ExcelIndex();
        int length = args3.IntValue();

        if (startIndex < 0) {
            return ParameterError(2);
        }
        if (length < 0) {
            return ParameterError(3);
        }
        if (startIndex >= text.length()) {
            return Operand.Create("");
        }
        if (startIndex + length > text.length()) {
            length = text.length() - startIndex;
        }
        return Operand.Create(text.substring(startIndex, startIndex + length));
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.TEXT;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.TEXT);
        func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func3.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
    }
}
